### PR TITLE
PG: join new vertex data by vertex ids

### LINKED ISSUE #2793: [BUG]Adding data to same nodes twice in a PG results in incorrect __vertex_prop_dataframe
**Describe the bug**
If i add data to the same nodes twice , we end up with incorrect number of rows and incorrect state of PG `__vertex_prop_dataframe` which causes a bunch of errors downstream.  

**Steps/Code to reproduce bug**
```python
import cugraph
from cugraph.experimental import PropertyGraph
import cudf

def add_node_a(pg):
    v_df =  cudf.DataFrame({'node_ids':[0,1,2]})
    pg.add_vertex_data(v_df, vertex_col_name='node_ids', type_name='a')

    feat_df =  cudf.DataFrame({'node_ids':[0,1,2], 'data':[10,20,30]})
    pg.add_vertex_data(feat_df, vertex_col_name='node_ids', type_name='a')

def add_node_b(pg):
    v_df =  cudf.DataFrame({'node_ids':[3,4,5]})
    pg.add_vertex_data(v_df, vertex_col_name='node_ids', type_name='b')

    feat_df =  cudf.DataFrame({'node_ids':[3,4,5], 'data':[10,20,30]})
    pg.add_vertex_data(feat_df, vertex_col_name='node_ids', type_name='b')
    
pg = PropertyGraph()
add_node_a(pg)
add_node_b(pg)
pg.get_num_vertices('b')
```
```
  | _TYPE_ | data
-- | -- | --
a | 10
a | 20
a | 30
b | <NA>
b | <NA>
b | <NA>
b | 10
b | 2
b | 30
```

```python
pg.get_num_vertices('b')
6
```
**Expected behavior**
```
3
```

**Environment details**
Please run and paste the output of the `cugraph/print_env.sh` script here, to gather any other relevant environment details

**Additional context**
This will probably block https://github.com/rapidsai/dgl/pull/16

CC: @eriknw , @rlratzel , @alexbarghi-nv , @BradReesWork

### LINKED ISSUE #2794: [BUG] Can not add data to the same edges twice
**Describe the bug**
Adding data to the same edges twice results in duplicated rows.  This prevents me adding features to an existing edge in PG. 
```python
import cugraph
import cudf
from cugraph.experimental import PropertyGraph


df_1 =  cudf.DataFrame({'src': [0,1],
                      'dst': [1,2],
                      'feat_1':[10,20]
                     })


df_2 =  cudf.DataFrame({'src': [0,1],
                       'dst': [1,2],
                       'feat_2':[0.5,0.2]
                     })

pg = PropertyGraph()
pg.add_edge_data(df_1, ['src','dst'], type_name="a")
pg.add_edge_data(df_2, ['src','dst'], type_name="a")
pg._EXPERIMENTAL__PropertyGraph__edge_prop_dataframe
```

```python
	_SRC_	_DST_	_TYPE_	feat_1	feat_2
_EDGE_ID_					
0	0	1	a	10	<NA>
1	1	2	a	20	<NA>
2	0	1	a	<NA>	0.5
3	1	2	a	<NA>	0.2
```

**Expected behavior**
I would expect that we add it to the existing rows.  

```python
	_SRC_	_DST_	_TYPE_	feat_1	feat_2
_EDGE_ID_					
0	0	1	a	10	0.5
1	1	2	a	20	0.2

```

CC: @eriknw , @rlratzel , @alexbarghi-nv , @BradReesWork

### PR DESCRIPTION
**Pull request by eriknw**
Fixes #2793
Fixes #2794

We are using the assumption that there should be a single row for each vertex id.

This does not yet handle MG.  Let's figure out how we want SG to behave first.

### CODE REVIEW

**Comment by eriknw** [comment]
This PR needs https://github.com/rapidsai/cudf/pull/11998 to pass tests.

I think we ought to try to benchmark this before merging. 

Here is the mini-benchmark `bench_add_edges_cyber` for "branch-22.12" branch (I added `[50]` for kicks):
```
------------------------------------------------------------- benchmark: 5 tests ------------------------------------------------------------
Name (time in s, mem in bytes)        Mean                GPU mem            GPU Leaked mem            Rounds            GPU Rounds
---------------------------------------------------------------------------------------------------------------------------------------------
bench_add_edges_cyber[1]            1.3833 (1.0)      273,342,072 (1.0)                 144 (1.0)           1           1
bench_add_edges_cyber[3]            2.8139 (2.03)     273,342,968 (1.00)              1,032 (7.17)          1           1
bench_add_edges_cyber[10]           6.7457 (4.88)     273,343,656 (1.00)              1,712 (11.89)         1           1
bench_add_edges_cyber[30]          18.8487 (13.63)    273,344,440 (1.00)              2,528 (17.56)         1           1
bench_add_edges_cyber[50]          30.2863 (21.89)    273,346,992 (1.00)              5,048 (35.06)         1           1
---------------------------------------------------------------------------------------------------------------------------------------------
```
and here is the benchmark for the current branch
```
--------------------------------------------------------------- benchmark: 5 tests ---------------------------------------------------------------
Name (time in ms, mem in bytes)            Mean                GPU mem            GPU Leaked mem            Rounds            GPU Rounds
--------------------------------------------------------------------------------------------------------------------------------------------------
bench_add_edges_cyber[1]               801.3284 (1.0)      273,342,224 (1.0)                 280 (1.0)           1           1
bench_add_edges_cyber[3]             1,723.6213 (2.15)     273,342,672 (1.00)                496 (1.77)          1           1
bench_add_edges_cyber[10]            4,731.0946 (5.90)     273,343,672 (1.00)              1,360 (4.86)          1           1
bench_add_edges_cyber[30]           13,608.0636 (16.98)    273,343,448 (1.00)              1,256 (4.49)          1           1
bench_add_edges_cyber[50]           22,819.4191 (28.48)    273,348,912 (1.00)              6,488 (23.17)         1           1
--------------------------------------------------------------------------------------------------------------------------------------------------
```
We shouldn't yet read too much into these numbers, but this PR actually isn't as fast as I was expecting. I'll investigate further and experiment with MAG240 dataset.

@rlratzel, this reworks the merge in `add_vertex_data` and `add_edge_data`, which seemed to be a particularly sensitive part of the code base. I think this PR can begin to be reviewed even though we're blocked on cudf PR for the tests.

File: python/cugraph/cugraph/dask/structure/mg_property_graph.py
Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/structure/mg_property_graph.py:458]: With `dask_cudf` merges/operations,  I think we should actually decrease number of partitions to  less than `_num_workers` unless we actually need to because it decrease parallelization and worker starving can become a problem.  

Though the shuffle cost of the merge is nlogn but if your input partitions are less than availble workers , worker starving can become a problem. 

Maybe lets pick `2*_num_workers` or something ?

Comment (VibhuJawa) [inline python/cugraph/cugraph/dask/structure/mg_property_graph.py:458]: With `dask_cudf` merges/operations,  I think we should actually decrease number of partitions to  less than `_num_workers` unless we actually need to because it decrease parallelization and worker starving can become a problem.  

Though the shuffle cost of the merge is `nlogn` but if your input partitions are less than available workers , worker starving can become a problem. 

Maybe lets pick `2*_num_workers` or something ?

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Minor clarification. Looks good other wise.

File: python/cugraph/cugraph/dask/structure/mg_property_graph.py
Comment (eriknw) [reply python/cugraph/cugraph/dask/structure/mg_property_graph.py]: That sounds reasonable. Done. I think we may want to explore these values as we scale out.

I know it's bad if we don't repartition though and the number of partitions grows too large.

Comment (VibhuJawa) [reply python/cugraph/cugraph/dask/structure/mg_property_graph.py]: Agreed . Yup ,  it becomes really slow as you increase partitions . In case you are curios on how the shuffle works under the hood, below is my favorite explanation by [rjzamora](https://github.com/rjzamora) which i think still holds true. 

https://github.com/rapidsai/cudf/pull/4308#issuecomment-598219871

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**
